src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/ReflectionXmlFormatReader.cs

```diff
@@ -88,33 +115,75 @@ private void ReflectionReadMembers(object obj, XmlReaderDelegator xmlReader, Xml
                 {
                     index = context.GetMemberIndex(xmlReader, memberNames, memberNamespaces, memberIndex, null);
                 }
-                ReflectionReadMember(obj, index, xmlReader, context, memberNames, memberNamespaces);
-                memberIndex = index;
-                requiredIndex = index + 1;
+
+                // GetMemberIndex returns memberNames.Length if member not found
+                if (index < members.Length)
+                {
+                    ReflectionReadMember(obj, index, xmlReader, context, members);
+                    memberIndex = index;
+                    requiredIndex = index + 1;
+                }
             }
         }
 
-        private int ReflectionReadMember(object obj, int memberIndex, XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, XmlDictionaryString[] memberNames, XmlDictionaryString[] memberNamespaces)
+        // Put all members of the contract (including types from base contract) into 'members'.
+        private int ReflectionGetMembers(ClassDataContract classContract, DataMember[] members)
         {
-            int memberCount = 0;
+            int memberCount = (classContract.BaseContract == null) ? 0 : ReflectionGetMembers(classContract.BaseContract, members);
+            int childElementIndex = memberCount;
+            for (int i = 0; i < classContract.Members.Count; i++, memberCount++)
+            {
+                members[childElementIndex + i] = classContract.Members[i];
+            }
+
+            return memberCount;
+        }
+
+        private void ReflectionReadMember(object obj, int memberIndex, XmlReaderDelegator xmlReader, XmlObjectSerializerReadContext context, DataMember[] members)
+        {
+            DataMember dataMember = members[memberIndex];
 
-            Type classType = _classContract.UnadaptedClassType;
-            int i = memberIndex + memberCount;
-            DataMember dataMember = _classContract.Members[i];
             Type memberType = dataMember.MemberType;
```

Comment:
Here is an assumption members[] will never contain a null.  Is that a valid assumption?  Possibly Contract.Assert(dataMember != null)